flac.go

```diff
@@ -253,7 +307,139 @@ func (stream *Stream) ParseNext() (f *frame.Frame, err error) {
 	return frame.Parse(stream.r)
 }
 
-// Seek is not implement yet.
-func (stream *Stream) Seek(offset int64, whence int) (read int64, err error) {
-	return 0, nil
+// Seek to a specific sample number in the flac stream.
+//
+// sample is valid if:
+// whence == io.SeekEnd and sample is negative
+// whence == io.SeekStart and sample is positive
+// whence == io.SeekCurrent and sample + current sample > 0 and < stream.Info.NSamples
+//
+// If sample does not match one of the above conditions then the result will
+// probably be seeking to the beginning or very end of the data and no error
+// will be returned.
+func (stream *Stream) Seek(sample int64, whence int) (read int64, err error) {
+	if stream.seekTable == nil && stream.seekTableSize > 0 {
+		if err := stream.makeSeekTable(); err != nil {
+			return 0, err
+		}
+	}
+
+	rs := stream.r.(io.ReadSeeker)
+
+	var point meta.SeekPoint
+	switch whence {
+	case io.SeekStart:
+		point = stream.searchFromStart(sample)
+	case io.SeekCurrent:
+		point, err = stream.searchFromCurrent(sample, rs)
+	case io.SeekEnd:
+		point = stream.searchFromEnd(sample)
+	default:
+		return 0, ErrInvalidSeek
+	}
+
+	if err != nil {
+		return 0, err
+	}
+
+	_, err = rs.Seek(stream.dataStart+int64(point.Offset), io.SeekStart)
+	return int64(point.SampleNum), err
+}
+
+func (stream *Stream) searchFromCurrent(sample int64, rs io.ReadSeeker) (p meta.SeekPoint, err error) {
+	o, err := rs.Seek(0, io.SeekCurrent)
+	if err != nil {
+		return p, err
+	}
+
+	offset := o - stream.dataStart
+	for _, p = range stream.seekTable.Points {
+		if int64(p.Offset) >= offset {
+			return stream.searchFromStart(int64(p.SampleNum) + sample), nil
+		}
+	}
+	return p, nil
+}
+
+// searchFromEnd expects sample to be negative.
+// If it is positive, it's ok, the last seek point will be returned.
+func (stream *Stream) searchFromEnd(sample int64) (p meta.SeekPoint) {
+	return stream.searchFromStart(int64(stream.Info.NSamples) + sample)
+}
+
+func (stream *Stream) searchFromStart(sample int64) (p meta.SeekPoint) {
```

Comment:
Rename `sample` to `sampleNum` for consistency with variable names used in `seektable.go`.